### PR TITLE
refactor(app): asynchronously load config state

### LINKED ISSUE #5869: refactor: app should load config into UI asyncronously
## overview

In order to unblock #5808, the app should load its configuration state asynchronously instead of synchronously.

The app currently depends on the configuration file being read and deserialized before the UI can be rendered. This is only possible because

1. The [configuration library](https://github.com/sindresorhus/electron-store) that we use does file IO synchronously
2. Electron provides a way to expose values to the UI at launch time.

If either of those puzzle pieces are removed, our config scheme doesn't work anymore. #5808 removes the (2) piece, and in general, it would be very cool to be unblocked from (1) if we ever needed to switch libraries or storage medium.

## implementation details

Right now, the app's config reducer state is pre-populated by a value that the app-shell `preload.js` script attaches to `window`. Instead:

- The `config` reducer should have an initial state of `null` ("unknown")
- The app-shell should send a `config:INITIALIZED` action to the UI when it hears that the UI is up

The possibility of `config` being `null` at app launch will bleed down to a lot of places in the app. Most, if not all, of those places should have sensible defaults to apply at the value level, but this has the potential to balloon the breadth of this work out a little

## acceptance criteria

This change must not noticeably affect the UX of the app

### PR DESCRIPTION
**Pull request by mcous**
## overview

This PR closes #5869. See ticket for background and acceptance criteria.

This is a big PR because a lot of things across the app were (understandably) reliant on the presence of config at all times, and on top of that, none of the config modification logic was covered by unit tets.

## changelog

- refactor(app): asynchronously load config state

This change exacerbated a problem with a config that we'd been punting on: certain config modifications (e.g. toggles) depend on the current value of the config. Previously, the UI components (or thunk actions) would have to read the previous value from `state.config` and construct the next value manually. However, if config isn't guaranteed to be in state, what is the UI supposed to do?

Rather than force the UI to deal with missing data and no-ops, this PR adds specific actions for these sorts of modifications (e.g. `config:TOGGLE_VALUE`). When these actions hit the app-shell, the config module is able to determine the best course of action.

## review requests

- [ ] App config toggles work
    - [ ] App config overrides via CLI and/or env. vars work and block toggles
- [ ] Intercom profile updates still function (profile initialization is now async)
- [ ] Mixpanel events still fire (analytics initialization is now async)
- [ ] Manual robot IP address list UI still functions
    - Note: in `make dev`, this config is overridden and you should _not_ be able to edit the list
    - To test editing the list, remove the `--discovery.candidates` option from the launch command in `app-shell/Makefile`

## risk assessment

Medium low risk. The app's config system is pretty vital. This PR addresses that risk by simplifying more complex config update operations (e.g. instead of forcing the UI to read a boolean value to toggle it, the UI can issue a toggle action) and increasing unit test coverage

### CODE REVIEW

File: app-shell/src/config/index.js [hunk 1]
```diff
@@ -70,14 +76,14 @@ export function getStore(): Config {
...
-  return getIn(overrides(), path)
+  return get(overrides(), path)
 }
 
 // TODO(mc, 2010-07-01): getConfig with path parameter can't be typed
 // Remove the path parameter
 export function getConfig(path?: string): any {
   const result = store().get(path)
-  const over = getIn(overrides(), path)
+  const over = get(overrides(), path)
```
Comment (mcous) [inline app-shell/src/config/index.js:86]: Note to self: Do not refactor stuff arbitrarily without unit tests. **Do not** refactor stuff arbitrarily without unit tests

(This breaks config overrides)

File: app/src/analytics/__tests__/epic.test.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
+// analytics epics tests
```
Comment (mcous) [inline app/src/analytics/__tests__/epic.test.js:1]: Mostly moved from `app/src/analytics/__tests__/epics.test.js`.

File: app/src/analytics/__tests__/epic.test.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      expect(actual).toEqual(expected)
+    })
+  })
+
+  afterEach(() => {
+    jest.resetAllMocks()
+  })
+
+  describe('initializeAnalyticsEpic', () => {
+    it('initializes analytics on config:INITIALIZED', () => {
```
Comment (mcous) [inline app/src/analytics/__tests__/epic.test.js:30]: New test

File: app/src/analytics/__tests__/epic.test.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        const state$ = hot('s-', { s: state })
+        const output$ = analyticsEpic(action$, state$)
+
+        expectObservable(output$).toBe('--')
+        flush()
+        expect(trackEvent).toHaveBeenCalledTimes(0)
+      })
+    })
+
+    it('handles events before config is initialized', () => {
```
Comment (mcous) [inline app/src/analytics/__tests__/epic.test.js:83]: New test

File: app/src/analytics/__tests__/epic.test.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        const action$ = hot('----')
+        const state$ = hot('-a-b', { a: off, b: off })
+        const output$ = analyticsEpic(action$, state$)
+
+        expectObservable(output$).toBe('----')
+        flush()
+        expect(setMixpanelTracking).toHaveBeenCalledTimes(0)
+      })
+
+      testScheduler.run(({ hot, expectObservable, flush }) => {
```
Comment (mcous) [inline app/src/analytics/__tests__/epic.test.js:150]: New expectation in existing test

File: app/src/analytics/epic.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+      trackEvent(event, config)
+    ),
+    ignoreElements()
+  )
+}
+
+const optIntoAnalyticsEpic: Epic = (_, state$) => {
+  return state$.pipe(
+    map<State, AnalyticsConfig | null>(
+      state => state.config?.analytics ?? null
```
Comment (b-cooper) [inline app/src/analytics/epic.js:54]: I see we're digging into top-level state to get to the config.analytics branch a couple times in this file. We often use selectors to abstract this digging out of places like this. Could we use a branch-level selector here instead?

**Comment by mcous** [comment]
> I had a thought about the UI of config toggles. Some of them may take a while between the UPDATE and SET actions, in this case, there is no feedback to the user that their click was registered, and this can lead to rage clicking, which will (correct me if I'm wrong) queue up all of those UPDATE actions

There's definitely a gap between the request and the response for changing **App** config in the strictest technical sense, but in a practical sense, I don't think there's any way for the delay between the request and the state update to be long enough to warrant an "in progress state". Since the request is going over IPC between two processes on the same machine (rather than, say, an HTTP request), the entire loop takes less than a hundredth of a second (on my machine, according to the Redux devtools)

**Pull request merged**